src/server/handlers/launchpad.js

```diff
@@ -0,0 +1,236 @@
+import { createHash } from 'crypto';
+
+import yaml from 'js-yaml';
+import Memcached from 'memcached';
+import parseGitHubUrl from 'parse-github-url';
+
+import { conf } from '../helpers/config';
+import requestGitHub from '../helpers/github';
+import getLaunchpad from '../launchpad';
+import logging from '../logging';
+
+const logger = logging.getLogger('express-error');
+
+// XXX cjwatson 2016-12-08: Hardcoded for now, but should eventually be
+// configurable.
+const DISTRIBUTION = 'ubuntu';
+const DISTRO_SERIES = 'xenial';
+const STORE_SERIES = '16';
+
+const RESPONSE_GITHUB_BAD_URL = {
+  status: 'error',
+  payload: {
+    code: 'github-bad-url',
+    message: 'Cannot parse GitHub URL'
+  }
+};
+
+const RESPONSE_GITHUB_NOT_FOUND = {
+  status: 'error',
+  payload: {
+    code: 'github-snapcraft-yaml-not-found',
+    message: 'Cannot find snapcraft.yaml in this GitHub repository'
+  }
+};
+
+const RESPONSE_GITHUB_AUTHENTICATION_FAILED = {
+  status: 'error',
+  payload: {
+    code: 'github-authentication-failed',
+    message: 'Authentication with GitHub failed'
+  }
+};
+
+const RESPONSE_GITHUB_OTHER = {
+  status: 'error',
+  payload: {
+    code: 'github-error-other',
+    message: 'Something went wrong when looking for snapcraft.yaml'
+  }
+};
+
+const RESPONSE_SNAPCRAFT_YAML_PARSE_FAILED = {
+  status: 'error',
+  payload: {
+    code: 'snapcraft-yaml-parse-failed',
+    message: 'Failed to parse snapcraft.yaml'
+  }
+};
+
+const RESPONSE_SNAPCRAFT_YAML_NO_NAME = {
+  status: 'error',
+  payload: {
+    code: 'snapcraft-yaml-no-name',
+    message: 'snapcraft.yaml has no top-level "name" attribute'
+  }
+};
+
+const RESPONSE_SNAP_NOT_FOUND = {
+  status: 'error',
+  payload: {
+    code: 'snap-not-found',
+    message: 'Cannot find existing snap based on this URL'
+  }
+};
+
+let memcached = null;
+
+const getMemcached = () => {
+  if (memcached === null) {
+    memcached = new Memcached(conf.get('MEMCACHED_HOST').split(','),
+                              { namespace: 'lp:' });
+  }
+  return memcached;
+};
+
+// Test affordance.
+export const setMemcached = (value) => {
+  memcached = value;
+};
+
+const makeSnapName = url => {
+  return createHash('md5').update(url).digest('hex');
+};
+
+const getSnapcraftYaml = (req, res, callback) => {
+  const repositoryUrl = req.body.repository_url;
+  const parsed = parseGitHubUrl(repositoryUrl);
+  if (parsed === null || parsed.owner === null || parsed.name === null) {
+    logger.info(`Cannot parse "${repositoryUrl}"`);
+    return res.status(400).send(RESPONSE_GITHUB_BAD_URL);
+  }
+
+  const uri = `/repos/${parsed.owner}/${parsed.name}/contents/snapcraft.yaml`;
+  const options = {
+    headers: {
+      'Authorization': `token ${req.session.token}`,
+      'Accept': 'application/vnd.github.v3.raw'
```

Comment:
Yes.  We could get it from the default JSON response, but it's more decoding for no gain.